pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,229 @@
+// 4. Wait for the context to be done
+// Note: this method is left empty on purpose and will be update with the actual implementation in the future PRs for the standalone DNS proxy
+func (s *FQDNDataServer) StreamPolicyState(stream pb.FQDNData_StreamPolicyStateServer) error {
+	// This is a temporary implementation to send the current state of the DNS rules to the client and used for testing
+	stream.Send(&pb.PolicyState{RequestId: "test"})
+	return nil
+}
+
+// NewServer creates a new FQDNDataServer which is used to handle the Standalone DNS Proxy grpc service
+func NewServer(endpointManager endpointmanager.EndpointManager, updateOnDNSMsg messagehandler.DNSMessageHandler, port int, logger *slog.Logger) *FQDNDataServer {
+	fqdnDataServer := &FQDNDataServer{
+		port:                port,
+		endpointManager:     endpointManager,
+		updateOnDNSMsg:      updateOnDNSMsg,
+		currentIdentityToIP: make(map[identity.NumericIdentity][]net.IPNet),
+		log:                 logger,
+		prefixLengths:       counter.DefaultPrefixLengthCounter(),
+	}
+
+	grpcServer := grpc.NewServer(grpc.KeepaliveEnforcementPolicy(kaep), grpc.KeepaliveParams(kasp))
+	fqdnDataServer.grpcServer = grpcServer
+	pb.RegisterFQDNDataServer(grpcServer, fqdnDataServer)
+	return fqdnDataServer
+}
+
+// OnIPIdentityCacheChange is a method to receive the IP identity cache change events
+func (s *FQDNDataServer) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr types.PrefixCluster, oldHostIP, newHostIP net.IP, oldID *ipcache.Identity, newID ipcache.Identity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata, endpointFlags uint8) {
+	s.identityToIPMutex.Lock()
+	defer s.identityToIPMutex.Unlock()
+
+	if cidr.ClusterID() != 0 {
+		return
+	}
+	ipNet := cidr.AsIPNet()
+	prefix := cidr.AsPrefix()
+	if cidr.ClusterID() == 0 {
+		switch modType {
+		case ipcache.Upsert:
+			if oldID != nil {
+				// Remove from the old identity
+				s.deleteFromIdentityToIPLocked(oldID, ipNet, prefix)
+			}
+			s.currentIdentityToIP[newID.ID] = append(s.currentIdentityToIP[newID.ID], ipNet)
+			s.prefixLengths.Add([]netip.Prefix{prefix})
+		case ipcache.Delete:
+			if oldID != nil {
+				s.deleteFromIdentityToIPLocked(oldID, ipNet, prefix)
+			}
+		}
+	}
+}
+
+// deleteFromIdentityToIPLocked deletes the given IP from the identity to IP mapping
+// It is called when the IP identity cache changes and the IP is deleted from the mapping
+// It is also called when the IP is upserted with a new identity
+// It removes the prefix from the prefixLengths map
+// It is called with the identityToIpMutex lock held
+func (s *FQDNDataServer) deleteFromIdentityToIPLocked(identity *ipcache.Identity, cidr net.IPNet, prefix netip.Prefix) error {
+	if identity == nil {
+		return fmt.Errorf("identity is nil")
+	}
+
+	if ips, ok := s.currentIdentityToIP[identity.ID]; ok {
+		newIPs := slices.DeleteFunc(ips, func(existing net.IPNet) bool {
+			if existing.String() == cidr.String() {
```

Comment:
Please use the `Equals` method here. Or switch to `netip.Prefix` everywhere as suggested above